das/daser.go

```diff
@@ -90,5 +148,97 @@ func (d *DASer) sampling(ctx context.Context, sub header.Subscription) {
+		// fully caught up to network head, it will resume DASing from this checkpoint
+		// up to current network head
+		// TODO @renaynay: Implement Share Cache #180 to ensure no duplicate DASing over same
+		//  header
+		if err := storeCheckpoint(d.cstore, checkpoint); err != nil {
+			log.Errorw("storing checkpoint to disk", "height", checkpoint, "err", err)
+		}
+		// signal that all catchUp routines have finished
+		close(d.catchUpDn)
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case job := <-d.jobsCh:
+			// perform catchUp routine
+			height, err := d.catchUp(ctx, job)
+			// update checkpoint before checking error as the height
+			// value from catchUp represents the last successfully DASed height,
+			// regardless of the routine's success
+			checkpoint = height
+			// exit routine if a catch-up job was unsuccessful
+			if err != nil {
+				log.Errorw("catch-up routine failed", "attempted range (from, to)", job.from,
+					job.to, "last successfully sampled height", height)
+				log.Warn("IN ORDER TO CONTINUE SAMPLING OVER PAST HEADERS, RE-START THE NODE")
+				return
+			}
+		}
+	}
+}
+
+// catchUp starts a sampling routine for headers starting at the next header
+// after the `from` height and exits the loop once `to` is reached. (from:to]
+func (d *DASer) catchUp(ctx context.Context, job *catchUpJob) (int64, error) {
+	routineStartTime := time.Now()
+	log.Infow("sampling past headers", "from", job.from, "to", job.to)
+
+	// start sampling from height at checkpoint+1 since the
+	// checkpoint height is DASed by broader sample routine
+	for height := job.from + 1; height <= job.to; height++ {
+		h, err := d.getter.GetByHeight(ctx, uint64(height))
+		if err != nil {
+			if err == context.Canceled {
+				// report previous height as the last successfully sampled height and
+				// error as nil since the routine was ordered to stop
+				return height - 1, nil
+			}
+
+			log.Errorw("failed to get next header", "height", height, "err", err)
+			// report previous height as the last successfully sampled height
+			return height - 1, err
+		}
+
+		startTime := time.Now()
+
+		err = d.da.SharesAvailable(ctx, h.DAH)
+		if err != nil {
+			if err == context.Canceled {
+				// report previous height as the last successfully sampled height and
+				// error as nil since the routine was ordered to stop
+				return height - 1, nil
+			}
+			log.Errorw("sampling failed", "height", h.Height, "hash", h.Hash(),
+				"square width", len(h.DAH.RowsRoots), "data root", h.DAH.Hash(), "err", err)
+			// continue sampling
+		}
+
+		sampleTime := time.Since(startTime)
+		log.Debugw("sampled past header", "height", h.Height, "hash", h.Hash(),
+			"square width", len(h.DAH.RowsRoots), "finished (s)", sampleTime.Seconds())
 	}
+
+	log.Infow("successfully caught up", "from", job.from,
+		"to", job.to, "finished (s)", time.Since(routineStartTime))
+	// report successful result
```

Comment:
log nit: a bit of incosistency with the starting log and success log. Should be something like:
`successfully sampled past headers`.

Reply:
resolved? https://github.com/celestiaorg/celestia-node/pull/473/commits/ad2975d10455ed92488bdbd81d5c30cf59e3a0bf